File: VB/DxSample/Filtering/ClientCriteriaVisitorBase.vb
Imports DevExpress.Data.Filtering

Namespace DxSample.Filtering

    Public Class ClientCriteriaVisitorBase
        Implements IClientCriteriaVisitor(Of CriteriaOperator)

        Protected Sub New()
        End Sub

        Protected Function AcceptOperator(ByVal theOperator As CriteriaOperator) As CriteriaOperator
            If ReferenceEquals(Nothing, theOperator) Then Return Nothing
            Return theOperator.Accept(Me)
        End Function

        Protected Function VisitOperands(ByVal operands As CriteriaOperatorCollection) As CriteriaOperatorCollection
            Dim isModified As Boolean = False
            Dim result As CriteriaOperatorCollection = New CriteriaOperatorCollection(operands.Count)
            For Each operand As CriteriaOperator In operands
                Dim acceptedOperand As CriteriaOperator = AcceptOperator(operand)
                result.Add(acceptedOperand)
                If Not ReferenceEquals(operand, acceptedOperand) Then isModified = True
            Next

            Return If(isModified, result, operands)
        End Function

        Protected Overridable Function VisitJoin(ByVal theOperand As JoinOperand) As CriteriaOperator
            Dim aggregatedExpression As CriteriaOperator = AcceptOperator(theOperand.AggregatedExpression)
            Dim condition As CriteriaOperator = AcceptOperator(theOperand.Condition)
            If ReferenceEquals(theOperand.AggregatedExpression, aggregatedExpression) AndAlso ReferenceEquals(theOperand.Condition, condition) Then Return theOperand
            Return New JoinOperand(theOperand.JoinTypeName, condition, theOperand.AggregateType, aggregatedExpression)
        End Function

        Protected Overridable Function VisitProperty(ByVal theOperand As OperandProperty) As CriteriaOperator
            Return theOperand
        End Function

        Protected Overridable Function VisitAggregate(ByVal theOperand As AggregateOperand) As CriteriaOperator
            Dim aggregatedExpression As CriteriaOperator = AcceptOperator(theOperand.AggregatedExpression)
            Dim condition As CriteriaOperator = AcceptOperator(theOperand.Condition)
            If ReferenceEquals(theOperand.AggregatedExpression, aggregatedExpression) AndAlso ReferenceEquals(theOperand.Condition, condition) Then Return theOperand
            Return New AggregateOperand(theOperand.CollectionProperty, aggregatedExpression, theOperand.AggregateType, condition)
        End Function

        Protected Overridable Function VisitFunction(ByVal theOperator As FunctionOperator) As CriteriaOperator
            Dim operands As CriteriaOperatorCollection = VisitOperands(theOperator.Operands)
            If ReferenceEquals(theOperator.Operands, operands) Then Return theOperator
            Return New FunctionOperator(theOperator.OperatorType, operands)
        End Function

        Protected Overridable Function VisitValue(ByVal theOperand As OperandValue) As CriteriaOperator
            Return theOperand
        End Function

        Protected Overridable Function VisitGroup(ByVal theOperator As GroupOperator) As CriteriaOperator
            Dim operands As CriteriaOperatorCollection = VisitOperands(theOperator.Operands)
            If ReferenceEquals(theOperator.Operands, operands) Then Return theOperator
            Return New GroupOperator(theOperator.OperatorType, operands)
        End Function

        Protected Overridable Function VisitIn(ByVal theOperator As InOperator) As CriteriaOperator
            Dim leftOperand As CriteriaOperator = AcceptOperator(theOperator.LeftOperand)
            Dim operands As CriteriaOperatorCollection = VisitOperands(theOperator.Operands)
            If ReferenceEquals(theOperator.LeftOperand, leftOperand) AndAlso ReferenceEquals(theOperator.Operands, operands) Then Return theOperator
            Return New InOperator(leftOperand, operands)
        End Function

        Protected Overridable Function VisitUnary(ByVal theOperator As UnaryOperator) As CriteriaOperator
            Dim operand As CriteriaOperator = AcceptOperator(theOperator.Operand)
            If ReferenceEquals(theOperator.Operand, operand) Then Return theOperator
            Return New UnaryOperator(theOperator.OperatorType, operand)
        End Function

        Protected Overridable Function VisitBinary(ByVal theOperator As BinaryOperator) As CriteriaOperator
            Dim leftOperand As CriteriaOperator = AcceptOperator(theOperator.LeftOperand)
            Dim rightOperand As CriteriaOperator = AcceptOperator(theOperator.RightOperand)
            If ReferenceEquals(theOperator.LeftOperand, leftOperand) AndAlso ReferenceEquals(theOperator.RightOperand, rightOperand) Then Return theOperator
            Return New BinaryOperator(leftOperand, rightOperand, theOperator.OperatorType)
        End Function

        Protected Overridable Function VisitBetween(ByVal theOperator As BetweenOperator) As CriteriaOperator
            Dim beginExpression As CriteriaOperator = AcceptOperator(theOperator.BeginExpression)
            Dim endExpression As CriteriaOperator = AcceptOperator(theOperator.EndExpression)
            Dim testExpression As CriteriaOperator = AcceptOperator(theOperator.TestExpression)
            If ReferenceEquals(theOperator.BeginExpression, beginExpression) AndAlso ReferenceEquals(theOperator.EndExpression, endExpression) AndAlso ReferenceEquals(theOperator.TestExpression, testExpression) Then Return theOperator
            Return New BetweenOperator(testExpression, beginExpression, endExpression)
        End Function

'#Region "IClientCriteriaVisitor"
        Private Function IClientCriteriaVisitor_Visit(ByVal theOperand As JoinOperand) As CriteriaOperator Implements IClientCriteriaVisitor(Of CriteriaOperator).Visit
            Return VisitJoin(theOperand)
        End Function

        Private Function IClientCriteriaVisitor_Visit1(ByVal theOperand As OperandProperty) As CriteriaOperator Implements IClientCriteriaVisitor(Of CriteriaOperator).Visit
            Return VisitProperty(theOperand)
        End Function

        Private Function IClientCriteriaVisitor_Visit2(ByVal theOperand As AggregateOperand) As CriteriaOperator Implements IClientCriteriaVisitor(Of CriteriaOperator).Visit
            Return VisitAggregate(theOperand)
        End Function

'#End Region
'#Region "ICriteriaVisitor"
        Private Function ICriteriaVisitor_Visit(ByVal theOperator As FunctionOperator) As CriteriaOperator Implements ICriteriaVisitor(Of CriteriaOperator).Visit
            Return VisitFunction(theOperator)
        End Function

        Private Function ICriteriaVisitor_Visit1(ByVal theOperand As OperandValue) As CriteriaOperator Implements ICriteriaVisitor(Of CriteriaOperator).Visit
            Return VisitValue(theOperand)
        End Function

        Private Function ICriteriaVisitor_Visit2(ByVal theOperator As GroupOperator) As CriteriaOperator Implements ICriteriaVisitor(Of CriteriaOperator).Visit
            Return VisitGroup(theOperator)
        End Function

        Private Function ICriteriaVisitor_Visit3(ByVal theOperator As InOperator) As CriteriaOperator Implements ICriteriaVisitor(Of CriteriaOperator).Visit
            Return VisitIn(theOperator)
        End Function

        Private Function ICriteriaVisitor_Visit4(ByVal theOperator As UnaryOperator) As CriteriaOperator Implements ICriteriaVisitor(Of CriteriaOperator).Visit
            Return VisitUnary(theOperator)
        End Function

        Private Function ICriteriaVisitor_Visit5(ByVal theOperator As BinaryOperator) As CriteriaOperator Implements ICriteriaVisitor(Of CriteriaOperator).Visit
            Return VisitBinary(theOperator)
        End Function

        Private Function ICriteriaVisitor_Visit6(ByVal theOperator As BetweenOperator) As CriteriaOperator Implements ICriteriaVisitor(Of CriteriaOperator).Visit
            Return VisitBetween(theOperator)
        End Function
'#End Region
    End Class
End Namespace

File: CS/DxSample/Filtering/ClientCriteriaVisitorBase.cs
using DevExpress.Data.Filtering;

namespace DxSample.Filtering {
    public class ClientCriteriaVisitorBase : IClientCriteriaVisitor<CriteriaOperator> {
        protected ClientCriteriaVisitorBase() { }

        protected CriteriaOperator AcceptOperator(CriteriaOperator theOperator) {
            if (object.ReferenceEquals(null, theOperator)) return null;
            return theOperator.Accept<CriteriaOperator>(this);
        }

        protected CriteriaOperatorCollection VisitOperands(CriteriaOperatorCollection operands) {
            bool isModified = false;
            CriteriaOperatorCollection result = new CriteriaOperatorCollection(operands.Count);
            foreach (CriteriaOperator operand in operands) {
                CriteriaOperator acceptedOperand = this.AcceptOperator(operand);
                result.Add(acceptedOperand);
                if (!object.ReferenceEquals(operand, acceptedOperand))
                    isModified = true;
            }
            return isModified ? result : operands;
        }

        protected virtual CriteriaOperator VisitJoin(JoinOperand theOperand) {
            CriteriaOperator aggregatedExpression = this.AcceptOperator(theOperand.AggregatedExpression);
            CriteriaOperator condition = this.AcceptOperator(theOperand.Condition);
            if (object.ReferenceEquals(theOperand.AggregatedExpression, aggregatedExpression) &&
                object.ReferenceEquals(theOperand.Condition, condition))
                return theOperand;
            return new JoinOperand(theOperand.JoinTypeName, condition, theOperand.AggregateType, aggregatedExpression);
        }

        protected virtual CriteriaOperator VisitProperty(OperandProperty theOperand) {
            return theOperand;
        }

        protected virtual CriteriaOperator VisitAggregate(AggregateOperand theOperand) {
            CriteriaOperator aggregatedExpression = this.AcceptOperator(theOperand.AggregatedExpression);
            CriteriaOperator condition = this.AcceptOperator(theOperand.Condition);
            if (object.ReferenceEquals(theOperand.AggregatedExpression, aggregatedExpression) &&
                object.ReferenceEquals(theOperand.Condition, condition))
                return theOperand;
            return new AggregateOperand(theOperand.CollectionProperty, aggregatedExpression, theOperand.AggregateType, condition);
        }

        protected virtual CriteriaOperator VisitFunction(FunctionOperator theOperator) {
            CriteriaOperatorCollection operands = this.VisitOperands(theOperator.Operands);
            if (object.ReferenceEquals(theOperator.Operands, operands))
                return theOperator;
            return new FunctionOperator(theOperator.OperatorType, operands);
        }

        protected virtual CriteriaOperator VisitValue(OperandValue theOperand) {
            return theOperand;
        }

        protected virtual CriteriaOperator VisitGroup(GroupOperator theOperator) {
            CriteriaOperatorCollection operands = this.VisitOperands(theOperator.Operands);
            if (object.ReferenceEquals(theOperator.Operands, operands))
                return theOperator;
            return new GroupOperator(theOperator.OperatorType, operands);
        }

        protected virtual CriteriaOperator VisitIn(InOperator theOperator) {
            CriteriaOperator leftOperand = this.AcceptOperator(theOperator.LeftOperand);
            CriteriaOperatorCollection operands = this.VisitOperands(theOperator.Operands);
            if (object.ReferenceEquals(theOperator.LeftOperand, leftOperand) &&
                object.ReferenceEquals(theOperator.Operands, operands))
                return theOperator;
            return new InOperator(leftOperand, operands);
        }

        protected virtual CriteriaOperator VisitUnary(UnaryOperator theOperator) {
            CriteriaOperator operand = this.AcceptOperator(theOperator.Operand);
            if (object.ReferenceEquals(theOperator.Operand, operand))
                return theOperator;
            return new UnaryOperator(theOperator.OperatorType, operand);
        }

        protected virtual CriteriaOperator VisitBinary(BinaryOperator theOperator) {
            CriteriaOperator leftOperand = this.AcceptOperator(theOperator.LeftOperand);
            CriteriaOperator rightOperand = this.AcceptOperator(theOperator.RightOperand);
            if (object.ReferenceEquals(theOperator.LeftOperand, leftOperand) &&
                object.ReferenceEquals(theOperator.RightOperand, rightOperand))
                return theOperator;
            return new BinaryOperator(leftOperand, rightOperand, theOperator.OperatorType);
        }

        protected virtual CriteriaOperator VisitBetween(BetweenOperator theOperator) {
            CriteriaOperator beginExpression = this.AcceptOperator(theOperator.BeginExpression);
            CriteriaOperator endExpression = this.AcceptOperator(theOperator.EndExpression);
            CriteriaOperator testExpression = this.AcceptOperator(theOperator.TestExpression);
            if (object.ReferenceEquals(theOperator.BeginExpression, beginExpression) &&
                object.ReferenceEquals(theOperator.EndExpression, endExpression) &&
                object.ReferenceEquals(theOperator.TestExpression, testExpression))
                return theOperator;
            return new BetweenOperator(testExpression, beginExpression, endExpression);
        }

        #region IClientCriteriaVisitor
        CriteriaOperator IClientCriteriaVisitor<CriteriaOperator>.Visit(JoinOperand theOperand) {
            return this.VisitJoin(theOperand);
        }

        CriteriaOperator IClientCriteriaVisitor<CriteriaOperator>.Visit(OperandProperty theOperand) {
            return this.VisitProperty(theOperand);
        }

        CriteriaOperator IClientCriteriaVisitor<CriteriaOperator>.Visit(AggregateOperand theOperand) {
            return this.VisitAggregate(theOperand);
        }
        #endregion
        #region ICriteriaVisitor
        CriteriaOperator ICriteriaVisitor<CriteriaOperator>.Visit(FunctionOperator theOperator) {
            return this.VisitFunction(theOperator);
        }

        CriteriaOperator ICriteriaVisitor<CriteriaOperator>.Visit(OperandValue theOperand) {
            return this.VisitValue(theOperand);
        }

        CriteriaOperator ICriteriaVisitor<CriteriaOperator>.Visit(GroupOperator theOperator) {
            return this.VisitGroup(theOperator);
        }

        CriteriaOperator ICriteriaVisitor<CriteriaOperator>.Visit(InOperator theOperator) {
            return this.VisitIn(theOperator);
        }

        CriteriaOperator ICriteriaVisitor<CriteriaOperator>.Visit(UnaryOperator theOperator) {
            return this.VisitUnary(theOperator);
        }

        CriteriaOperator ICriteriaVisitor<CriteriaOperator>.Visit(BinaryOperator theOperator) {
            return this.VisitBinary(theOperator);
        }

        CriteriaOperator ICriteriaVisitor<CriteriaOperator>.Visit(BetweenOperator theOperator) {
            return this.VisitBetween(theOperator);
        }
        #endregion
    }
}
